_posts/2019-09-20-lambda-to-openfaas.md

```diff
@@ -0,0 +1,282 @@
+```shell
+$ faas-cli secret create shorturl-dynamo-key --from-file=./access-key-id
+$ faas-cli secret create shorturl-dynamo-secret --from-file=./secret-access-key
+```
+
+To access these secrets in the function, add the following lines:
+```javascript
+let accessKeyId = fs.readFileSync("/var/openfaas/secrets/shorturl-dynamo-key").toString()
+let secretKey = fs.readFileSync("/var/openfaas/secrets/shorturl-dynamo-secret").toString()
+```
+
+Pass them into the AWS SDK configuration so that the DynamoDB `DocumentClient` will have access to the table:
+```javascript
+AWS.config.update({
+    region: 'us-west-1',
+    credentials: {
+        accessKeyId: accessKeyId,
+        secretAccessKey: secretKey
+    }
+});
+```
+Finally, we'll add the secrets to the function's `yml` file so that the function will be given the appropriate permissions to read the secret values
+```yaml
+functions:
+  shortener:
+    lang: node10-express
+    ...
+    secrets:
+      - shorturl-dynamo-key
+      - shorturl-dynamo-secret
+```
+
+> Read more about [Unifying Secrets with OpenFaaS](/blog/unified-secrets)
+
+#### Code Differences
+Below is a picture of what the differences for the "shortener" function look like as seen from a `git diff` (here, using VSCode)
+
+![Function Code Diff](/images/lambda-to-openfaas/code-diff.png)
+
+- We added the `fs` module to be able to read the secrets
+- Removed the `JSON.parse()` call, as the input is already in JSON format, so we are able to read the properties directly
+- Call a new function `configureAWS()` before creating the new `DynamoDB.DocumentClient`
+- Created a new function `configureAWS()` that reads the secrets, and sets the credentials and region for AWS
+- Replaced the AWS specific response object with the template's `context`
+- Set the `context.success` and `context.fail` status codes and values
+
+If you look at the `git diff --stat` output, we actually removed more code than we added!
+
+```shell
+ 1 file changed, 35 insertions(+), 37 deletions(-)
+```
+
+Also, notice that the AWS credentials are not in the code directly, so this function is still safe to commit to any repository without fear of leaking your credentials.
+
+#### Build and Deploy
+Now that our function code is migrated, it's time to build and deploy it!
+
+Both of these can be accomplished with the `faas-cli` tool. Optionally, with [OpenFaaS Cloud](https://docs.openfaas.com/openfaas-cloud/), this can all happen automatically when the changes are pushed to Github or GitLab.
+
+First, let's combine the two generated function `.yml` files so that both can be built and deployed at the same time.
+```shell
+$ mv shortener.yml stack.yml
+```
+
+Now, copy the `redirector` properties from the `redirecter.yml` file, and paste them in below the `shortener` function property.
```

Comment:
What about the append flag?

Reply:
Good idea! 

I moved this piece up to the "create a function" section adding the `--append` flag to the steps.